New architecture migration

louiszawadzki opened the pull request:
### What does this PR do?

Uses the new architecture for native modules.

### How to read this PR

This PR is quite big and has a lot of file changes. 
However, the changes are quite similar for all 4 modules.

My recommandation would be to go commit by commit, and read each module separately inside commits.

There is documentation on how to achieve the migration and examples as well:
- examples: https://github.com/react-native-community/RNNewArchitectureLibraries
- doc for the JS part: https://reactnative.dev/docs/new-architecture-library-intro
- doc for Android: https://reactnative.dev/docs/new-architecture-library-android
- doc for iOS: https://reactnative.dev/docs/new-architecture-library-ios

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)
- [ ] If this PR is auto-generated, please make sure also to manually update the code related to the change

Commit:
Add native modules specs

Commit:
Modify native code for new arch

Commit:
Add Android code for new and old archs

Commit:
Add Codegen config

Commit:
Use new arch in js

Commit:
Make native modules extend specs

louiszawadzki commented on packages/core/android/detekt.yml:
```diff
@@ -329,7 +329,7 @@ naming:
     ignoreOverridden: true
   InvalidPackageDeclaration:
-    active: true
+    active: false
```
Here I'm disabling this rule to match a bit more the "traditional" RN architecture and to avoid having too many nested folders (in `main`, `newarch` and `oldarch`). 
As we have to also make sure the path are correct in the sourceSets, I believe this reduces a bit the chances of making a mistake there.

Let me know if you think we should instead keep this rule.

0xnm replied:
Can you show what is the typical issue reported by this rule?

louiszawadzki commented on packages/core/android/src/oldarch/src/DdLogs.kt:
```diff
@@ -0,0 +1,163 @@
+    reactContext: ReactApplicationContext
+) : ReactContextBaseJavaModule(reactContext) {
+
+    private val implementation = DdLogsImplementation()
```
Question: here I'm not using `lazy` to initialize the implementation, because I think the `lazy` init of the `Logger` inside `DdLogsImplementation` is enough, is it correct?

0xnm replied:
depends on the need why `lazy` was put in the first place. Was it placed to avoid having NoOp implementation?

louiszawadzki replied:
Yes exactly

0xnm replied:
should be enough then I believe

Commit:
Migrate DdTrace to turbomodules for iOS

Commit:
Migrate DdLogs to turbomodules for iOS

Commit:
Migrate DdRum to turbomodules for iOS

Commit:
Migrate DdSdk to turbomodules for iOS

Commit:
Fix Swift function name

Commit:
Add file otherwise missing from Swift header file

Commit:
Add missing copyright header in new files

Commit:
Keep QueueHolder file in Swift

Commit:
Only define new arch if enabled

Commit:
Remove .only in tests

Commit:
Use the new architecture in webview

maxep commented on packages/core/ios/Sources/DdLogs.mm:
```diff
@@ -0,0 +1,149 @@
+
+RCT_EXPORT_MODULE()
+
+RCT_REMAP_METHOD(debug, withDebugMessage:(NSString*)message
+                 withContext:(NSDictionary*)context
+                 withResolver:(RCTPromiseResolveBlock)resolve
+                 withRejecter:(RCTPromiseRejectBlock)reject)
+{
+    [self debug:message context:context resolve:resolve reject:reject];
+}
```
Do you need 2 layers here?
I think you can do this directly:
```suggestion
RCT_REMAP_METHOD(debug, withDebugMessage:(NSString*)message
                 withContext:(NSDictionary*)context
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)
{
    [self.ddLogsImplementation debugWithMessage:message context:context resolve:resolve reject:reject];
}
```

no?

louiszawadzki replied:
I will look into it!

louiszawadzki replied:
I looked into it and we can indeed build without having this duplication.
However, this duplication is what enables us to ensure the module matches the definition of the JS Spec, `RCT_REMAP_METHOD` does seem to include the method in the properties for now.

So it's a question of balance:
- Benefit: we get easy type checking when modifying an API + we are sure of the type checking
- Drawback: duplicate code whenever we add a new function

For me it's worth keeping it as it is now, RN might change in the future so we don't have to do this duplicate declaration and we can always remove it if it's too much pain to maintain in the end.

maxep commented on packages/core/ios/Sources/DdLogs.h:
```diff
@@ -0,0 +1,24 @@
+
+#ifdef RCT_NEW_ARCH_ENABLED
+
+#import <DdSdkReactNative/DdSdkReactNative.h>
```
This import can be removed as `DdSdkReactNative.h` is blank 👍

louiszawadzki replied:
Actually `DdSdkReactNative/DdSdkReactNative.h` is generated when running `pod install` but not committed. It contains the interfaces for the native module, derived from the JS spec. That's why it's blank here :)

louiszawadzki replied:
I'll add some docs on how to develop the SDK in the new architecture though, including how to generate this file

maxep commented on packages/core/ios/Sources/DdLogs.h:
```diff
@@ -0,0 +1,24 @@
+
+#endif
+
+@property (nonatomic, strong) DdLogsImplementation* ddLogsImplementation;
```
I might be wrong, but I don't think you need to expose this in the `.h`, it can stay private to the class.

It's also the case in all declaration, `DdRum`, `DdTrace`, ...

louiszawadzki replied:
I guess I could add an extension to the interface like described here: https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ObjectiveC/Chapters/ocCategories.html#//apple_ref/doc/uid/TP30001163-CH20-SW2

To me it's a nice to have, the impact is minimal and the change might create confusion.

maxep reviewed (approved):
Well done! I left few suggestions but it looks great overall.

On a side note: Prefixes in the RN bridge always bugged me 😄  If you want a more idiomatic namings, you should use:
`DDLogs.h`
`DDlogs.mm`
`Logs.swift`

Not a change request but we have a good opportunity here 🙂

Commit:
Add example app for new architecture

Commit:
Initialize SDK in new arch app

Commit:
Add documentation for building for the new RN arch

Commit:
Integrate new arch repo with workflows

Commit:
Specify that number arguments are non null

Not specifying it triggers an error like this one: https://github.com/facebook/react-native/issues/27920.
I believe this error was not thrown before as we did not use NSNumber in the previous implementation, but NSInteger.

fuzzybinary commented on example-new-architecture/App.tsx:
```diff
@@ -0,0 +1,155 @@
+ * Sample React Native App
+ * https://github.com/facebook/react-native
+ *
+ * @format
```
Out of curiosity, What's this `@format` for?

Also, should this have the license comment at the top instead?

louiszawadzki replied:
That's probably to tell some linter tool to format this file by default.
We don't have headers in the `example` app, we don't ship this code as part of the SDK and it's technically been generated, so not sure we must have license header on these ones.

0xnm commented on example-new-architecture/android/app/build.gradle:
```diff
@@ -0,0 +1,170 @@
+            storePassword 'android'
+            keyAlias 'androiddebugkey'
+            keyPassword 'android'
+        }
```
I believe this is not needed. By default AGP will generate default keystore anyway, no need to have a common one for everyone.

Also having keystore + credentials in the repo will be flagged by various security scanners and it will take time to mark these issues as FP and explain them.

0xnm commented on example-new-architecture/android/app/debug.keystore:
No need to commit this file, AGP is able to generate keystore for debug configuration on the fly.

0xnm commented on example-new-architecture/App.tsx:
```diff
@@ -0,0 +1,155 @@
+  setTimeout(() => {
+    DdRum.addTiming('one_second');
+  }, 1000);
+  await DdRum.addAction(RumActionType.CUSTOM, 'custom action');
+  DdLogs.info('info log');
```
why some calls to our SDK are with `await` and some are without it? Isn't they all return `Promise`?

louiszawadzki replied:
I've added `await` only for those who need to be resolved before making the next one, but I can `await` all of them actually

0xnm replied:
do we explain to the customer when they need to add `await` and when not with our SDK? or it is not needed?

louiszawadzki replied:
I don't think that it needs to be explained, it very much depends on the use case (if you want a log to be associated with a trace or an action for instance).
I think our users are able to figure this out by themselves, I haven't seen any support case around this so far

0xnm commented on example-new-architecture/android/app/src/main/java/com/ddsdkreactnativeexample/MainApplication.java:
```diff
@@ -0,0 +1,62 @@
+
+        @Override
+        protected boolean isNewArchEnabled() {
+          return BuildConfig.IS_NEW_ARCHITECTURE_ENABLED;
```
if this app specifically targets new architecture (while `example` targets the old one), shouldn't we return always `true` here? Or we are making a possibility for this app to use old architecture as well?

louiszawadzki replied:
Doing so indeed keeps the possibility to use the old architecture, but that's not the main reason for keeping it as it is.
Changing the generated Java files might make it harder to upgrade React Native in the future, and we also keep it closer to how React Native apps enable the new architecture currently (using `gradle.properties`) so it's easy for any React Native dev to understand how this works.

0xnm commented on example-new-architecture/android/app/src/main/res/values/strings.xml:
```diff
@@ -0,0 +1,3 @@
+<resources>
+    <string name="app_name">DdSdkReactNativeExample</string>
```
minor: we can give more human-friendly name

0xnm commented on example-new-architecture/android/build.gradle:
```diff
@@ -0,0 +1,21 @@
+
+buildscript {
+    ext {
+        buildToolsVersion = "33.0.0"
+        minSdkVersion = 21
+        compileSdkVersion = 33
+        targetSdkVersion = 33
```
btw release `34` is already available (we didn't update from 33 to 34 in the Android SDK yet though)

louiszawadzki replied:
If that's ok I will keep it to 33 to match the RN app template (I don't know if RN is compatible with 34 already)

0xnm replied:
totally, this was just a note from me for the future.

fuzzybinary commented on packages/core/DatadogSDKReactNative.podspec:
```diff
@@ -23,4 +23,17 @@ Pod::Spec.new do |s|
   end
+
+  
+  # This guard prevent to install the dependencies when we run `pod install` in the old architecture.
```
```suggestion
  # This guard prevents installing the dependencies when we run `pod install` in the old architecture.
```

0xnm commented on example-new-architecture/app.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "DdSdkReactNativeExample",
+  "displayName": "DdSdkReactNativeExample"
```
can be something more human-readable maybe?

fuzzybinary commented on packages/core/ios/Sources/DdLogs.mm:
```diff
@@ -0,0 +1,149 @@
+- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
+    (const facebook::react::ObjCTurboModule::InitParams &)params
+{
+    return std::make_shared<facebook::react::NativeDdLogsSpecJSI>(params);
```
Does react automatically generate these `SpecJSI` structures?  Hence why they're in the `facebook::react` namespace?

louiszawadzki replied:
Exactly :)

fuzzybinary reviewed (commented):
LGTM, just a few random questions

0xnm commented on packages/core/android/src/main/kotlin/src/DdSdkReactNativePackage.kt:
```diff
@@ -0,0 +1,78 @@
+ */
+class DdSdkReactNativePackage : TurboReactPackage() {
+    override fun getModule(name: String, reactContext: ReactApplicationContext): NativeModule? {
+        return if (name == DdSdkImplementation.NAME) {
```
better to convert to `when(name)`

0xnm commented on packages/core/android/src/main/kotlin/src/DdSdkReactNativePackage.kt:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Package of native dd-sdk-reactnative native modules.
+ */
+class DdSdkReactNativePackage : TurboReactPackage() {
```
So this is using `TurboReactPackage` which comes from the new architecture, but the class itself sits in the `main` source set, and not in `newarch`? this won't cause problems?

louiszawadzki replied:
This won't cause problems as there is retro compatibility for the old architecture in TurboReactPackage (see the `isTurboModule` variables further down when registering the packages).

0xnm commented on packages/core/android/src/main/kotlin/src/DdSdkReactNativePackage.kt:
```diff
@@ -0,0 +1,78 @@
+    override fun getReactModuleInfoProvider(): ReactModuleInfoProvider {
+        return ReactModuleInfoProvider {
+            val moduleInfos: MutableMap<String, ReactModuleInfo> =
+                HashMap()
```
probably can be just the following:

```suggestion
            val moduleInfos = mutableMapOf<String, ReactModuleInfo>()
```

0xnm commented on packages/core/android/src/main/kotlin/src/DdSdkReactNativePackage.kt:
```diff
@@ -0,0 +1,78 @@
+                true,  // hasConstants
+                false,  // isCxxModule
+                isTurboModule // isTurboModule
+            )
```
I would rather simplify it to the following given that params for `ReactModuleInfo` are always the same (except the name). And maybe mutable collection above is not needed at all.

```suggestion
            val moduleInfos = listOf(
              DdSdkImplementation.NAME,
              DdRumImplementation.NAME,
              DdTraceImplementation.NAME,
              DdLogsImplementation.NAME
            ).associateWith {
              ReactModuleInfo(
                it,
                it,
                false,  // canOverrideExistingModule
                false,  // needsEagerInit
                true,  // hasConstants
                false,  // isCxxModule
                isTurboModule // isTurboModule
              )
            }
```

0xnm commented on packages/core/android/src/main/kotlin/src/DdTraceImplementation.kt:
why did many files moved to the root of the source sets, not respecting the package structure? Seems like only rename to `DdTrace` to `DdTraceImplementation` was needed, but not the move?

0xnm commented on packages/core/android/src/newarch/src/DdLogs.kt:
```diff
@@ -0,0 +1,148 @@
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.reactnative
```
file location doesn't respect the package name. It should be placed in the `com/datadog/reactnative` folder, same for other files.

0xnm commented on packages/core/android/src/newarch/src/DdLogs.kt:
```diff
@@ -0,0 +1,148 @@
+/**
+ * The entry point to use Datadog's Logs feature.
+ */
+class DdLogs(
```
suggestion: maybe it is worth to name this class `DdLogsSpec` or `DdLogsModule` and keep the old class `DdLogs` without renaming to `DdLogsImplementation`? wdyt? not important point for me though, just for the consistency with the base class naming.

louiszawadzki replied:
That's a fair point. 
I'd see one reason to keep it as it is now, it's that it respects the naming in all the new architecture examples. That would help anyone looking at this to understand which file is doing what more easily (and we can also maybe have the SDK showcased in migrations ;)).
So I'd be in favour of keeping the naming as it is for now, and maybe switch later when the new architecture is more widely spread, what do you think?

0xnm replied:
I'm fine with that, I just didn't enjoy having `Implementation` suffix, but I guess we can live with that.

0xnm commented on packages/core/android/src/oldarch/src/DdLogs.kt:
```diff
@@ -0,0 +1,163 @@
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.reactnative
```
in this source set all the files should also respect packaging structure and should be placed under `com/datadog/reactnative` folder.

0xnm commented on packages/core/src/logs/DdLogs.ts:
```diff
@@ -39,7 +37,9 @@ const isLogWithError = (
 class DdLogsWrapper implements DdLogsType {
-    private nativeLogs: DdNativeLogsType = NativeModules.DdLogs;
+    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
+    private nativeLogs: DdNativeLogsType = require('../specs/NativeDdLogs')
```
if we are compiling for the old architecture (because for Kotlin we have 2 source sets), will it still work?

louiszawadzki replied:
Yes as TurboModules work for both old and new arch (the internal packages of React Native have been using TurboModules since version `0.62` at least, even though they always ran on the old arch)

0xnm commented on packages/core/src/specs/NativeDdLogs.ts:
```diff
@@ -0,0 +1,104 @@
+/**
+ * Do not import this Spec directly, use DdNativeLogsType instead.
+ */
+export interface Spec extends TurboModule {
```
such files are generated I guess? judging by the missing copyright.

0xnm reviewed (commented):
nice work! I left some comments/questions. Please note, that all Kotlin files should still reside under `com/datadog/reactnative` folder to respect package naming (currently this is not respected).

0xnm commented on example-new-architecture/android/app/build.gradle:
```diff
@@ -0,0 +1,164 @@
+    buildTypes {
+        debug {
+            signingConfig signingConfigs.debug
+        }
```
can remove empty stuff and `debug` build type by default points to `debug` signing config anyway:

```suggestion
    buildTypes {
```

0xnm commented on example-new-architecture/android/app/src/main/java/com/ddsdkreactnativeexample/MainActivity.java:
```diff
@@ -0,0 +1,35 @@
+package com.ddsdkreactnativeexample;
```
maybe it makes sense to use `com.datadog.reactnativeexample` (or just `com.datadog.reactnative`) package (and folder structure) for the sample app.

louiszawadzki replied:
I reused the same package names between the 2 example apps to facilitate signing (especially on iOS, so you don't have to create 2 app ids and provisioning profiles).
Not sure this is critical to change as it's just a sample application?

0xnm replied:
it is ok for me to have it like that, but this is Android app and package name can be different from application ID, it doesn't affect signing.

0xnm commented on packages/core/android/build.gradle:
```diff
@@ -81,15 +89,20 @@ android {
+      if (isNewArchitectureEnabled()) {
+        java.srcDirs += ['src/newarch']
+      } else {
+        java.srcDirs += ['src/oldarch']
```
for the naming consistency with the `main` source set it makes sense to add `kotlin` folder as well, like `src/newarch/kotlin` and move files under it in these sets.

0xnm reviewed (approved):
lgtm! I've added few suggestions

Commit:
Use public instead of open in Swift
